### PR TITLE
Don't count core classes as free electives NP-37

### PR DESCRIPTION
**Pull request by cubetastic33**
NP-37

Only leftover courses that aren't used in core or major should be counted as free electives.

### CODE REVIEW

File: validator/degree_solver.py [hunk 1]
```diff
@@ -1,9 +1,11 @@
 from __future__ import annotations
 from enum import Enum
-import json
+from copy import deepcopy
 from pydantic import Json
 
 from typing import Any
+
+import major.requirements.shared
```
Comment (akevinge) [inline validator/degree_solver.py:8]: Remove unused imports.

File: validator/degree_solver.py [hunk 2]
```diff
@@ -172,12 +174,19 @@ def solve(self) -> DegreeRequirementsSolver:
             self.solved_core = core_solver.solve(
                 [course for course in self.courses], []
             )  # Convert to list
+        # Set of the core courses that are fulfilled, so they won't be considered as free electives
+        core_courses = set()
```
Comment (akevinge) [inline validator/degree_solver.py:178]: Nit: naming could be more explicit? `used_core_courses` or something along those lines might be better.

**Review by akevinge — DISMISSED** [review]
As discussed offline, major requirements (i.e. free electives) should still accept partially unused core courses due to hour splitting. This implemented causes all used core courses to be omitted from the major solver - even if they have unused hours.  

We can make another PR for that since this one fixes the immediate issue.

Also, some nitpicks. PTAL.

File: validator/degree_solver.py (hunk 1 above)
Comment (cubetastic33) [reply validator/degree_solver.py]: good catch

File: validator/degree_solver.py (hunk 2 above)
Comment (cubetastic33) [reply validator/degree_solver.py]: makes sense

**Review by akevinge — APPROVED** [review]
Edit: PTAL

File: validator/degree_solver.py [hunk 3]
```diff
@@ -1,9 +1,10 @@
 from __future__ import annotations
 from enum import Enum
-import json
 from pydantic import Json
 
 from typing import Any
+
+import major.requirements.shared
```
Comment (akevinge) [inline validator/degree_solver.py:7]: This is also unused.

**Review by akevinge — APPROVED** [review]

**Review by akevinge — APPROVED** [review]

**Pull request merged**